File: jung-algorithms/src/test/java/edu/uci/ics/jung/algorithms/shortestpath/TestUnweightedShortestPath.java
/*
 * Created on Aug 22, 2003
 *
 */
package edu.uci.ics.jung.algorithms.shortestpath;

import com.google.common.collect.BiMap;
import com.google.common.graph.GraphBuilder;
import com.google.common.graph.MutableGraph;
import edu.uci.ics.jung.algorithms.util.Indexer;
import java.util.function.Supplier;
import junit.framework.Assert;
import junit.framework.Test;
import junit.framework.TestCase;
import junit.framework.TestSuite;

/**
 * @author Scott White
 */
public class TestUnweightedShortestPath extends TestCase {
  private Supplier<String> nodeFactory =
      new Supplier<String>() {
        int count = 0;

        public String get() {
          return "V" + count++;
        }
      };

  BiMap<String, Integer> id;

  @Override
  protected void setUp() {}

  public static Test suite() {
    return new TestSuite(TestUnweightedShortestPath.class);
  }

  public void testUndirected() {
    MutableGraph<String> ug = GraphBuilder.undirected().allowsSelfLoops(true).build();
    for (int i = 0; i < 5; i++) {
      ug.addNode(nodeFactory.get());
    }
    id = Indexer.<String>create(ug.nodes());

    //		GraphUtils.addNodes(ug,5);
    //		Indexer id = Indexer.getIndexer(ug);
    ug.putEdge(id.inverse().get(0), id.inverse().get(1));
    ug.putEdge(id.inverse().get(1), id.inverse().get(2));
    ug.putEdge(id.inverse().get(2), id.inverse().get(3));
    ug.putEdge(id.inverse().get(0), id.inverse().get(4));
    ug.putEdge(id.inverse().get(4), id.inverse().get(3));

    UnweightedShortestPath<String> usp = new UnweightedShortestPath<String>(ug);
    Assert.assertEquals(usp.getDistance(id.inverse().get(0), id.inverse().get(3)).intValue(), 2);
    Assert.assertEquals(
        (usp.getDistanceMap(id.inverse().get(0)).get(id.inverse().get(3))).intValue(), 2);
    Assert.assertNull(usp.getIncomingEdgeMap(id.inverse().get(0)).get(id.inverse().get(0)));
    Assert.assertNotNull(usp.getIncomingEdgeMap(id.inverse().get(0)).get(id.inverse().get(3)));
  }

  public void testDirected() {
    MutableGraph<String> dg = GraphBuilder.directed().allowsSelfLoops(true).build();
    for (int i = 0; i < 5; i++) {
      dg.addNode(nodeFactory.get());
    }
    id = Indexer.<String>create(dg.nodes());
    dg.putEdge(id.inverse().get(0), id.inverse().get(1));
    dg.putEdge(id.inverse().get(1), id.inverse().get(2));
    dg.putEdge(id.inverse().get(2), id.inverse().get(3));
    dg.putEdge(id.inverse().get(0), id.inverse().get(4));
    dg.putEdge(id.inverse().get(4), id.inverse().get(3));
    dg.putEdge(id.inverse().get(3), id.inverse().get(0));

    UnweightedShortestPath<String> usp = new UnweightedShortestPath<String>(dg);
    Assert.assertEquals(usp.getDistance(id.inverse().get(0), id.inverse().get(3)).intValue(), 2);
    Assert.assertEquals(
        (usp.getDistanceMap(id.inverse().get(0)).get(id.inverse().get(3))).intValue(), 2);
    Assert.assertNull(usp.getIncomingEdgeMap(id.inverse().get(0)).get(id.inverse().get(0)));
    Assert.assertNotNull(usp.getIncomingEdgeMap(id.inverse().get(0)).get(id.inverse().get(3)));
  }
}

File: jung-algorithms/src/test/kotlin/edu/uci/ics/jung/algorithms/shortestpath/TestUnweightedShortestPath.kt
/*
 * Created on Aug 22, 2003
 *
 */
package edu.uci.ics.jung.algorithms.shortestpath

import com.google.common.collect.BiMap
import com.google.common.graph.GraphBuilder
import edu.uci.ics.jung.algorithms.util.Indexer
import junit.framework.Assert
import junit.framework.Test
import junit.framework.TestCase
import junit.framework.TestSuite
import java.util.function.Supplier

/**
 * @author Scott White
 */
class TestUnweightedShortestPath : TestCase() {
  private val nodeFactory = object : Supplier<String> {
    var count = 0
    override fun get(): String {
      return "V" + count++
    }
  }

  lateinit var id: BiMap<String, Int>

  override fun setUp() {}

  companion object {
    fun suite(): Test {
      return TestSuite(TestUnweightedShortestPath::class.java)
    }
  }

  fun testUndirected() {
    val ug = GraphBuilder.undirected().allowsSelfLoops(true).build<String>()
    for (i in 0 until 5) {
      ug.addNode(nodeFactory.get())
    }
    id = Indexer.create<String>(ug.nodes())

    ug.putEdge(id.inverse()[0]!!, id.inverse()[1]!!)
    ug.putEdge(id.inverse()[1]!!, id.inverse()[2]!!)
    ug.putEdge(id.inverse()[2]!!, id.inverse()[3]!!)
    ug.putEdge(id.inverse()[0]!!, id.inverse()[4]!!)
    ug.putEdge(id.inverse()[4]!!, id.inverse()[3]!!)

    val usp = UnweightedShortestPath<String>(ug)
    Assert.assertEquals(usp.getDistance(id.inverse()[0]!!, id.inverse()[3]!!)!!, 2)
    Assert.assertEquals(
        (usp.getDistanceMap(id.inverse()[0]!!)[id.inverse()[3]!!])!!, 2)
    Assert.assertNull(usp.getIncomingEdgeMap(id.inverse()[0]!!)[id.inverse()[0]!!])
    Assert.assertNotNull(usp.getIncomingEdgeMap(id.inverse()[0]!!)[id.inverse()[3]!!])
  }

  fun testDirected() {
    val dg = GraphBuilder.directed().allowsSelfLoops(true).build<String>()
    for (i in 0 until 5) {
      dg.addNode(nodeFactory.get())
    }
    id = Indexer.create<String>(dg.nodes())
    dg.putEdge(id.inverse()[0]!!, id.inverse()[1]!!)
    dg.putEdge(id.inverse()[1]!!, id.inverse()[2]!!)
    dg.putEdge(id.inverse()[2]!!, id.inverse()[3]!!)
    dg.putEdge(id.inverse()[0]!!, id.inverse()[4]!!)
    dg.putEdge(id.inverse()[4]!!, id.inverse()[3]!!)
    dg.putEdge(id.inverse()[3]!!, id.inverse()[0]!!)

    val usp = UnweightedShortestPath<String>(dg)
    Assert.assertEquals(usp.getDistance(id.inverse()[0]!!, id.inverse()[3]!!)!!, 2)
    Assert.assertEquals(
        (usp.getDistanceMap(id.inverse()[0]!!)[id.inverse()[3]!!])!!, 2)
    Assert.assertNull(usp.getIncomingEdgeMap(id.inverse()[0]!!)[id.inverse()[0]!!])
    Assert.assertNotNull(usp.getIncomingEdgeMap(id.inverse()[0]!!)[id.inverse()[3]!!])
  }
}
